### ISSUE TITLE
Cannot listen to the messages posted by users who hasn't installed the app

### ISSUE THREAD

**Issue by johngrigoryan** [slackapi/bolt-python#723]
Hello! 
I am developing a slack app with python, which should work on user behalf. The event listener should retrieve all the messages, from the workspace's channels and direct messages that the user, who installed the app, has access to. To achieve that goal I am using Events API with Oauth authorization and user scopes.

#### The `slack_bolt` version

slack-bolt==1.14.3
slack-sdk==3.18.3
slackeventsapi==3.0.1

#### Python runtime version

Python 3.8.10

#### Steps to reproduce:
1. Here is my Oauth settings and App setup.
```python
oauth_settings = OAuthSettings(
    client_id=os.environ["SLACK_CLIENT_ID"],
    client_secret=os.environ["SLACK_CLIENT_SECRET"],
    user_scopes=user_scopes,
    installation_store=FileInstallationStore(base_dir="./data/installations"),
    state_store=FileOAuthStateStore(expiration_seconds=600, base_dir="./data/states")
)

app = App(signing_secret=os.environ["SLACK_SIGNING_SECRET"], oauth_settings=oauth_settings)

api = FastAPI()
app_handler = SlackRequestHandler(app)
```
2. You can see that I am using FastAPI to handle the requests. Here is the FastAPI setup.
```python
@api.get("/slack/oauth_redirect")
async def oauth_redirect(req: Request):
    await app_handler.handle(req)
    return logging.info('Installation completed.')


@api.get("/slack/install")
async def install(req: Request):
    return await app_handler.handle(req)


@api.post("/slack/events")
async def endpoint(req: Request):
    return await app_handler.handle(req)
```
3. And this is the example test event listener function, that should just print out in the console the payload of the message.
```python
@app.event('message')
def listener(payload, client):
    print(payload)
```
### Expected result:

I am expecting to see every message's payload in the console posted in all the channels and direct messages, where the user is added (in this stage I am the user who installed the app to the workspace, also I am the admin of the workspace where the app is installed to).

### Actual result:

Every time that the message is posted by a user, who hasn't installed the app, I am getting this message in my console:
![Screenshot from 2022-09-23 14-45-53](https://user-images.githubusercontent.com/57317379/191944853-2a33c8de-efe0-4906-8f0a-de65757bcaeb.png)

Thanks in advance!

**Comment by hello-ashleyintech** [comment]
Hi, @johngrigoryan! Thanks for your question!

First glancing at the code provided, since Bolt for Python will provide the needed routes (such as `/slack/install`) when using OAuth, I'm not sure if the usage of the `api` variable and consequently `FastAPI` is needed here. The routes being provided from Bolt is [default behavior of `OAuthSettings`](https://github.com/slackapi/bolt-python/blob/b30a319ee848a6f2faaa6a83774f50704b6217c6/slack_bolt/oauth/oauth_settings.py#L90-L92) - you can find more info in the second and third paragraphs under [this section](https://slack.dev/bolt-python/concepts#authenticating-oauth) of the docs. However, let me know if I'm missing some context regarding that for your specific use case and we can troubleshoot further from there!

Regarding the error message - for the scenario you described, since you have OAuth enabled, the user who is installing the app to the workspace will need to install it through the `/slack/install` endpoint. If it is deemed that the use of `api` is not needed here, I would recommend removing that portion and the request handling, installing the app as the user through the `/slack/install` endpoint, and then trying to activate that message event listener. 

Additionally, if you are trying to log your event payload, you could update your listener so that it pulls the information from `event`:
```js
@app.event('message')
def listener(event):
    print(event)
```

Let me know if you have any additional questions, would be happy to help further! 🙌

**Comment by johngrigoryan** [comment]
Hi Ashley, thank you for the answer.

About the request handling that the api variable does. If I would remove that part from the code, I will get this error message when I would try to connect to the /slack/install path. `{"detail":"Not Found"}`. So, the need of the FastAPI is to handle the requests coming from slack server to the ngrok webserver.

I already installed the app through the `/slack/install` endpoint.

Also, this part didn't work
```python
@app.event('message')
def listener(event):
    print(event)
```

**Comment by hello-ashleyintech** [comment]
@johngrigoryan - gotcha! Thanks for that clarification. To clarify, have you added in your Redirect URL for `ngrok` as specified [here](https://github.com/slack-samples/bolt-python-starter-template#app-distribution--oauth) in your app configuration settings under OAuth and Permissions? This might also help offset the need for the FastAPI, but let me know if that is something you've already done or if that doesn't suit your particular use case!

In terms of the message listener - are you still getting the same error or a different one?

**Comment by johngrigoryan** [comment]
Thank you for getting back to me. Yes, I added the url from `ngrok` to the Redirect URL. I am still getting the same error message. What I think is that can be because `InstallationStore` or `InstallationStoreAuthorize`.

**Comment by hello-ashleyintech** [comment]
@johngrigoryan Based [on this comment](https://github.com/slackapi/bolt-python/blob/main/slack_bolt/middleware/authorization/multi_teams_authorization.py#L73-L75), this situation will either arise in the following scenarios:
```js
                # * A developer installed the app from the "Install to Workspace" button in Slack app config page
                # * The InstallationStore failed to save or deleted the installation for this workspace
```
Since you're using a file installation store, have you checked to see if the user installation instance was logged in your data store and all information has been populated and logged properly? If that installation instance has not been populated, it could be an issue of how the installation and OAuth requests are being called or handled.

**Comment by hello-ashleyintech** [comment]
@johngrigoryan are you using any frameworks for your Python app, such as Django?

**Comment by johngrigoryan** [comment]
thank you very much. only FastAPI.

**Comment by hello-ashleyintech** [comment]
@johngrigoryan I have some additional ideas:

1. The first one is to adjust the `/slack/oauth_redirect` request handler to return the request response rather than the logging:
```python
@api.get("/slack/oauth_redirect")
async def oauth_redirect(req: Request):
    return await app_handler.handle(req)
```

Once updating this, I would uninstall or delete the installation instance(s) of the app in your datastore that is returning the error, then reinstall and try the event listener and see if the behavior has changed. There is a chance because the request response was not returned, the installation may not have been completed successfully.

2. If that doesn't work, you can try to reworking your app configuration so it aligns more with the FastAPI OAuth example [here](https://github.com/slackapi/bolt-python/blob/main/examples/fastapi/oauth_app.py).

This would mean adjusting it so that there is nothing being passed into your `App()` initialization, including omitting the `OAuthSettings`, and letting the FastAPI installation requests and environment variables handle the installation settings. You should also still keep the `/slack/oauth_redirect` request so that it returns the request response rather than the logging.

With reworking what you have, the API and app initialization portion of your configuration would likely look similar to this:
```python
## Import needed things up here 

# App() will pull the output of os.environ.get("SIGNING_SECRET") by default if not set: https://github.com/slackapi/bolt-python/blob/b30a319ee848a6f2faaa6a83774f50704b6217c6/slack_bolt/app/app.py#L183
# App() will pull the output of os.environ.get("SLACK_BOT_TOKEN") by default if not set: https://github.com/slackapi/bolt-python/blob/b30a319ee848a6f2faaa6a83774f50704b6217c6/slack_bolt/app/app.py#L184
app = App()
app_handler = SlackRequestHandler(app)

@app.event('message')
def listener(payload, client):
    print(payload)

api = FastAPI()

@api.get("/slack/oauth_redirect")
async def oauth_redirect(req: Request):
    return await app_handler.handle(req)


@api.get("/slack/install")
async def install(req: Request):
    return await app_handler.handle(req)


@api.post("/slack/events")
async def endpoint(req: Request):
    return await app_handler.handle(req)

# # -- OAuth flow -- #
export SLACK_SIGNING_SECRET=os.environ["SLACK_SIGNING_SECRET"]
export SLACK_CLIENT_ID=os.environ["SLACK_CLIENT_ID"]
export SLACK_CLIENT_SECRET=os.environ["SLACK_CLIENT_SECRET"]
export SLACK_SCOPES=app_mentions:read,channels:history,im:history,chat:write # update with your scopes

## run app
```

Let me know if the above or something like it works for you! If you run into any additional errors or have any other information that you think might be helpful to troubleshoot, feel free to drop it here and I can take a look!

**Comment by johngrigoryan** [comment]
Hi Aubrey, sorry for the delay.

I tried the 1st scenario you mentioned, but, unfortunately it didn't work for me. Also, could you please explain the 2nd point more detailed, I really don't understand how the things're going to work if I would delete the Oauth part from my code.

Thank you for your time and patience.

**Comment by johngrigoryan** [comment]
Also, what I managed to notice, it seems that Slack bolt API tries to get authorization token for any user who sends a message to the user, who installed the app. Maybe, I am not right, but otherwise I can't find an explanation for the problem. Could you please tell me when and why the MultiTeamsAuthorization middleware should be triggered? The `Although the app should be installed into this workspace, the AuthorizeResult (returned value from authorize) for it was not found.` error message is written in the code of that middleware.

Thanks.

**Comment by johngrigoryan** [comment]
Would like to mention that I activated public distribution too, maybe that can help somehow.

**Comment by hello-ashleyintech** [comment]
@johngrigoryan, thanks for the updates! After doing a little more research, I wonder if this error is coming up for users who don’t have the app installed due to a token issue. As stated [here](https://github.com/slackapi/bolt-python/issues/695#issuecomment-1207202775), it seems like if your app requires user token scopes, then individual users would need to install the app to fulfill this requirement, whereas if your app needs only bot token scopes, then only one person in a workspace needs to install the app.

I recommend looking at what scopes you’re trying to pass in to your app in the App config and also seeing what scopes are added to your app in the app settings page - if there are any user token scopes added or being passed into your app config that you do not need, I would recommend removing them. For this specific scenario with the message listener, it sounds like you’d only need Bot token scopes. Otherwise, if you are requiring user token scopes that you need for the app functionality and a specific user does not have the app installed as you described originally, this seems like expected behavior.

**Comment by johngrigoryan** [comment]
Thanks for getting back to me.
I am using user tokens in my app, because my goal is to get all the messages, where the current user has access to(the user who authorized to the app). As mentioned in the Slack API's documentation bot scopes don't match my case because to achieve my goal with using the bot scopes I will need to add the app to the conversations from where I would like to grab the messages. Before testing the app on another workspace, it worked perfect on the workspace where I installed it first, so I guess that it is possible to do on multiple workspaces.
To be more clear, let me give an example.
I have 2 users, user A and user B. The user A has installed app in the workspace. When the user B sends a message to user A the event listener should extract some data from the payload of the message. This part worked properly before oauth installation setting and public distribution activation. Now it seems that Bolt tries to authorize the user B to the app.

**Comment by hello-ashleyintech** [comment]
@johngrigoryan gotcha! thanks for adding that clarification! 🙌 

Out of curiosity, does the same error get thrown if User A sends a message to User B or is it able to listen to it successfully?

**Comment by johngrigoryan** [comment]
In the case when user A sends a message to user B it listens successfully.

**Comment by hello-ashleyintech** [comment]
@johngrigoryan thanks so much for checking on that for me! I've checked internally to see if anyone has run into a similar error before and it looks like we might need more info to get a clearer picture of what's happening.

Could you send the following:
* A list of all [event types](https://api.slack.com/events) that you're listening to for your app, including `message`
* The payload that is being sent when User B messages User A and the error is thrown
* A log of the event for the use case where the error is thrown
* Your app manifest (you can find this in your App Settings under the Features heading - you can send it as JSON format)

All of this information will be helpful for us to troubleshoot this further! 🙌

**Comment by seratch** [comment]
Hi @johngrigoryan, it seems that you are trying to use only user tokens and don't want to have a bot token. That's the reason why you are facing the confusion here.

[The built-in InstallationStoreAuthorize](https://github.com/slackapi/bolt-python/blob/v1.14.3/slack_bolt/authorization/authorize.py#L103-L306) tries to resolve at least one token for an incoming request. This is a by-design behavior of Bolt framework, so you cannot change this. 

When your app don't have a bot token and has only a user token for User A, **the authorize function can resolve token only for the user A**. [Bolt intentionally removes User A's user token from authorize result for requests associated with any other users](https://github.com/slackapi/bolt-python/blob/v1.14.3/slack_bolt/authorization/authorize.py#L184-L189) to avoid unexpected data exposure to others (e.g., displaying messages in private channels/DMs, posting a message as User A in response to User B's actions / behaviors).

So, for your use case, there are two options:
1. Add bot scopes such as commands and issue a bot token for installations
2. Write your own authorize function, which always returns User A's token if it works for your use case

I highly recommend going with the option 1. Either way, you will need to have some bot scopes for other features such as interactivity, Home tab, and so on. Also, having a bot user in installed workspaces can help end-users easily find your app.

I hope this was helpful to you!

**Comment by johngrigoryan** [comment]
Hi @seratch thank you for the answer.

If I would add some bot tokens, will I be able to properly use the app on multiple workspaces? Also, please look at [the comment above](https://github.com/slackapi/bolt-python/issues/723#issuecomment-1258248528) where I tried to describe the goal of the app.

Thanks.

**Comment by seratch** [comment]
@johngrigoryan Your InstallationStore can manage bot tokens for multiple workspaces. I would suggest seeing how [the default file installation store](https://github.com/slackapi/python-slack-sdk/blob/v3.18.3/slack_sdk/oauth/installation_store/file/__init__.py) saves installation data in your local file system. With the default settings, it stores installation data under `$HOME/.bolt-app-installation`. In the same way, any of [the built-in installation stores](https://github.com/slackapi/python-slack-sdk/tree/main/slack_sdk/oauth/installation_store) and your own ones can properly manage installations for multiple workspaces.

Also, your app can subscribe both bot events and user events (either is fine too). So, as long as you have **necessary events on behalf of users** in the app settings, your app can receive events specific to installed users.

Is everything clear?

**Comment by johngrigoryan** [comment]
@seratch so if I add some bot scopes and generate a bot token, will the event client use that bot token? Supposing a user who hasn't installed the app sends a message to a user who has authorized/installed the app, will the slack web API use the bot token or the user token to get information about the conversation in which the message is posted?
Am I right that in case of using web API client with bot token I will not be able to get conversation info without adding the app to the conversation?

**Comment by seratch** [comment]
@johngrigoryan 
>so if I add some bot scopes and generate a bot token, will the event client use that bot token? 

As for the client argument of your listeners, yes, it will. That being said, you can still overwrite it with `context.user_token` if it exists. You can pass the user token as token argument for client's method calls (e.g., `client.chat_postMessage(token=context.user_token, ...)`).

Before answering the next question, let me explain how the event delivery works in details first. As for the user token use cases, please note that the story is not so simple. In a nutshell, you cannot expect event deliveries per user. Let's say User A and User B installed your app in a workspace and they are in the same channel. In this case, every single message event from the channel can be delivered **only once, not twice**. This means that Slack Platform does not duplicate the same event delivery per user. Thus, if you have multiple on-behalf-of-user event subscriptions, you may need to use https://api.slack.com/methods/apps.event.authorizations.list to know the full list of the users associated with a single event delivery.

>Supposing a user who hasn't installed the app sends a message to a user who has authorized/installed the app, will the slack web API use the bot token or the user token to get information about the conversation in which the message is posted?

As I mentioned above, you can overwrite the token by passing a different one as a method argument named "token". But more importantly, your app may need to determine which user token (or tokens) to use for the Web API calls in the way I've mentioned above (= fetching the full list of associated installing users of an event).

>Am I right that in case of using web API client with bot token I will not be able to get conversation info without adding the app to the conversation?

Yes, you are right. So, when your app receives a message event from a channel where your bot user is not a member of, your app needs to use a relevant user's user token for some web API callls.

I know that all the things I mentioned here are complicated but I hope my answers make sense to you.

**Comment by johngrigoryan** [comment]
Thank you for getting back to me @seratch.

> Thus, if you have multiple on-behalf-of-user event subscriptions, you may need to use https://api.slack.com/methods/apps.event.authorizations.list to know the full list of the users associated with a single event delivery.

So, using this method I will be able to find out to which users the event is related? Do I need app level token to use it? And if yes, where should I pass the token in my code?

Many thanks for your quick answers and patience, I really appreciate this.

**Comment by seratch** [comment]
@johngrigoryan 
>So, using this method I will be able to find out to which users the event is related? Do I need app level token to use it? And if yes, where should I pass the token in my code?

Yes for all the questions.

**Comment by johngrigoryan** [comment]
@seratch one last question, how to use both bot token and app-level token in my app?

**Comment by seratch** [comment]
@johngrigoryan You can issue a new app-level token in the app admin page. And then, you can add the value as a global variable, which can be loaded from environment variables or some similar ways. Your bot/user tokens per workspace are managed by your `InstallationStore`. Bolt resolves them for each incoming request in runtime.

```python
import os
app_level_token_client = WebClient(token=os.get("SLACK_APP_LEVEL_TOKEN"))

@app.event("message")
def handle_message_events(context, client, body):
    # `context` can have a bot token (always exists) and a user token (may exist)
    # `client` is associated to context.bot_token
    # These are resolved per a request

    # You can use a global app-level token wired client too
    # You can do this in a custom global middleware too
    auth_list_response = app_level_token_client.apps_event_authorizations_list(body["event_context"])
    # You may want to fetch user tokens based on auth_list_response["authorizations"]
    user_tokens = fetch_user_tokens(auth_list_response["authorizations"])
    # ....
```

**Comment by johngrigoryan** [comment]
@seratch @hello-ashleyintech Thank you very much guys!
